diff --git a/src/Nix/Pretty.hs b/src/Nix/Pretty.hs
index 6801923..5f8b498 100644
--- a/src/Nix/Pretty.hs
+++ b/src/Nix/Pretty.hs
@@ -120,6 +120,12 @@ wrapPath op sub =
     (dquotes $ antiquote sub)
     (wasPath sub)
 
+prettyReceiver :: NixDoc ann -> Doc ann
+prettyReceiver sub =
+  if rootOp sub == appOpDef
+    then parens (getDoc sub)
+    else wrap appOpDef sub
+
 -- | Handle Output representation of the string escape codes.
 prettyString :: NString (NixDoc ann) -> Doc ann
 prettyString (DoubleQuoted parts) =
@@ -288,7 +294,7 @@ exprFNixDoc = \case
       (mkNixDoc selectOp)
       (const leastPrecedence)
       o
-      $ wrapPath selectOp (mkNixDoc selectOp (wrap appOpDef r')) <> "." <> prettySelector attr <>
+      $ wrapPath selectOp (mkNixDoc selectOp (prettyReceiver r')) <> "." <> prettySelector attr <>
         ((" or " <>) . precedenceWrap appOpDef) `whenJust` o
    where
     selectOp :: NOperatorDef
